rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/SpatialMatcher.scala

```diff
@@ -264,119 +248,130 @@ object SpatialMatcher extends SpatialMatcherInstances {
-            // We also succeed if there isn't but the remainder is empty.
-            if (wildcard || rem == Nil)
-              NonDetFreeMapWithCost.pure(Unit)
-            else
-              // This should be prevented by the length checks.
-              NonDetFreeMapWithCost.emptyMap[Unit]
-          // If there's a capture variable, we prefer to add things to that rather than throw them
-          // away.
-          case Some(level) => {
-            handleRemainder(rem, level, merger, wildcard)
-          }
-        }
-      // Try to find a match for a single pattern.
-      case (targets, pattern +: prem) => {
-        if (!lf.connectiveUsed(pattern)) {
-          possiblyRemove(pattern, targets) match {
-            case None => NonDetFreeMapWithCost.liftF(Stream.empty)
-            case Some(filtered) =>
-              listMatch(filtered, prem, merger, varLevel, wildcard)
-          }
-        } else {
-          for {
-            trem        <- listMatchItem(targets, pattern, spatialMatch[T, T])
-            forcedYield <- listMatch(trem, prem, merger, varLevel, wildcard)
-          } yield forcedYield
-        }
+  def listMatchSingle[T](tlist: Seq[T], plist: Seq[T])(
+      implicit lf: HasLocallyFree[T],
+      sm: SpatialMatcher[T, T]): OptionalFreeMapWithCost[Unit] =
+    listMatchSingleNonDet(tlist, plist, (p: Par, _: T) => p, None, false)
+      .mapK[OptionT[State[CostAccount, ?], ?]](
+        new FunctionK[StreamT[State[CostAccount, ?], ?], OptionT[State[CostAccount, ?], ?]] {
+          override def apply[A](
+              fa: StreamT[State[CostAccount, ?], A]): OptionT[State[CostAccount, ?], A] =
+            OptionT(fa.value.map(_.headOption))
+        })
+
+  private[this] def listMatch[T](tlist: Seq[T],
+                                 plist: Seq[T],
+                                 merger: (Par, T) => Par,
+                                 varLevel: Option[Int],
+                                 wildcard: Boolean)(
+      implicit lf: HasLocallyFree[T],
+      sm: SpatialMatcher[T, T]): OptionalFreeMapWithCost[Unit] = {
+
+    type Pattern = (T, Stream[OptionalFreeMapWithCost[T]])
+
+    //TODO enforce sorted-ness of targets and patterns.
+    //Correctness of the matching relies on them being sorted.
+    //Use a type class? A type tag?
+    val termsSorted    = tlist
+    val patternsSorted = plist.zip(Stream.continually(termsSorted))
+
+    def guard(predicate: => Boolean): OptionalFreeMapWithCost[Unit] =
+      if (predicate) OptionalFreeMapWithCost.pure(()) else OptionalFreeMapWithCost.emptyMap
+
+    val maximumBipartiteMatch = MaximumBipartiteMatch((p: T, t: T) => {
+      val matchEffect = if (!lf.connectiveUsed(p)) {
+        guard(t == p).modifyCost(_.charge(COMPARISON_COST))
+      } else {
+        spatialMatch(t, p)
       }
-    }
+      matchEffect.attempt.map(_.isRight)
+    })
+
+    for {
+      matchedTargetsOpt <- maximumBipartiteMatch.findMatches(patternsSorted.toList)
```

Comment:
There's no way that this is correct for remainders anyway.
For wildcards you're fine, but for remainders you need #(patterns) - #(targets) pseudo patterns that only match if locallyFree is empty.